README.md
# foundry-vtt-expiry
![Foundry VTT Expiry](artwork/module/expiry-cover.png)

## Summary
The idea of food spoilage in Foundry VTT has been inquired about before; macros and other such exist as proof-of-concepts for interested GMs, and there may be a system or two with mechanics around it.

Expiry is a module that seeks to give GMs unlimited flexibility in creating food that automatically spoils and changes state within the player's inventory. A spoiled apple can rot away entirely or become a different consumable, one with the chance to poison an unlucky player.

Expiry makes use of the community-standard Simple Calendar module to track time in-game.

CHANGELOG.md
# foundry-vtt-expiry

## 1.0.2
- Configuration dialog improvements
- Basic logic implementation for item name changes

## 1.0.1
- Initial POC code

scratchpad.js
Hooks.on("createItem", function(document, _, userId) {
    if(!document.flags?.expiry) {
        console.log("item not configured for expiry")
        return false;
    } else {
        console.log("item configured for expiry")
        const newExpiryObject = new expiryObject(document, _, userId)
        newExpiryObject.createExpiryObject();
    };
    Hooks.off("createItem");
});

Hooks.on("simple-calendar-date-time-change", function(data) {
    const expiryActors = getExpiryActors()
    for(const expiryActor of expiryActors) {
        const affectedItems = getRelevantInventory(expiryActor);
        for(const affectedItem of affectedItems) {
            console.log("updating expiry item")
            const newExpiryObject = new expiryObject(affectedItem, data);
            newExpiryObject.updateExpiryObject();
        };
    };
    Hooks.off("simple-calendar-date-time-change");
});

class expiryObject {
    constructor(objectDocument, updateData=null, userId=null) {
        this.objectDocument = objectDocument;
        this.updateData = updateData;
        this.userId = userId;
        this.currentTime = SimpleCalendar.api.timestamp()
    };

    createExpiryObject() {
        let parentExpiry = this.objectDocument.flags.expiry;
        this.objectDocument.update({
            flags: {
                expiry: {
                    instance: {
                        createTime: this.currentTime,
                        createName: this.objectDocument.name,
                        activeStage: 0,
                    }
                }
            }
        });
        console.log("created expiry item")
        Hooks.call('createdExpiry', [this.objectDocument, this.userId])
    };

    checkExpiry() {
        if(!this.objectDocument.flags?.expiry?.enabled == true) {
            return false;
        };

        if(this.currentTime >= this.objectDocument.flags.expiry.instance.createTime) {
            this.updateExpiry();
        } else {
            return false;
        };
    };
    
    updateExpiryObject() {
        let expiryData = this.objectDocument.flags.expiry;
        
        // Calculate the time difference in seconds since creation
        const timePassed = this.currentTime - expiryData.instance.createTime;

        // Calculate the number of decay intervals that have passed
        const decaySteps = Math.floor(timePassed / expiryData.decayInterval);

        // Get the array of decay stages from the parentExpiry
        const decayStages = Object.keys(expiryData.decayStages);

        // Determine the current decay stage
        const stageIndex = Math.min(decaySteps, decayStages.length - 1);

        // Update the active stage
        if(!expiryData.instance.activeStage !== stageIndex){
            this.objectDocument.update({
                flags: {
                    expiry: {
                        instance: {
                            activeStage: stageIndex
                        }
                    }
                }
            });
            this._decayExpiryObject(decayStages[stageIndex]);
        };
    };

    _decayExpiryObject(decayStage) {
        // apply decay effects, emit a hook, what have you
        console.log(`decaying item ${this.objectDocument.id}`)
        decayAction = decayStage.action
        if(decayAction.revise){
            this.objectDocument.update({
                img: decayAction.revise.img || this.objectDocument.img,
                name: `${decayAction.revise.prefix}${this.objectDocument.flags.expiry.instance.createName}${decayAction.replace.suffix}` || this.objectDocument.name
            });
        } else if(decayAction.replace){
            const parentActor = objectDocument.parent;
            const replacementItem = game.items.get(decayAction.replace.itemId);
            if(await parentActor.createEmbeddedDocuments('Item', [replacementItem.toObject()])){
                this.objectDocument.delete();
            } else {
                console.error(`unable to create replacement object for Expiry ${this.objectDocument.id}, leaving parent object in-place`);
            }
        } else if(decayStage.action.remove){
            this.objectDocument.delete();
        };

        if(decayStage.message){
            ui.notifications.warn(decayState.message);
        };

    };
};

function getRelevantInventory(actorDocument) {
    const affectedItems = actorDocument.items.filter(item => {
      const expiryFlags = item.flags?.expiry;
      if (!expiryFlags) {
        // If the item doesn't have expiry flags, skip it
        return false;
      }
  
      const currentTime = SimpleCalendar.api.timestamp(); // Get current in-game time
      const instance = expiryFlags.instance;
      const decayStages = expiryFlags.decayStages;
  
      if (!instance || !decayStages) {
        // If either instance or decayStages is missing, skip the item
        return false;
      }
  
      // Calculate the entire lifespan of the item
      const lifespanEnd =
        instance.createTime +
        (Object.keys(decayStages).length * expiryFlags.decayInterval);
  
      // Check if the item's createTime is before or equal to the current time,
      // and if the item is not in its final decay state OR if the entire lifespan
      // is greater than or equal to the current time
      return (
        instance.createTime <= currentTime &&
        (!isInFinalDecayState(instance.activeStage, decayStages) ||
          lifespanEnd >= currentTime)
      );
    });
  
    return affectedItems;
  }
  
  function isInFinalDecayState(activeStage, decayStages) {
    const stages = Object.keys(decayStages);
    return activeStage === stages.length - 1;
  }

function getExpiryActors() {
    // Gets a list of actors who hold expiry-configured items
    return game.actors.filter(actor => actor.items.filter(item => item.flags.expiry));
};

let parentSettings = {
    flags: {
        expiry: {
            decayInterval: 86400,
            decayStages: {
                fresh: {
                    actions: {
                        revise: {
                            rename: {  // What name changes should be made e.g. `Fresh <item name>` or `<item name> (Middling)`
                                suffix: null,
                                prefix: null
                            },
                            reimage: {  // What new item picture should be used
                                newImage: texturePath
                            },
                        },
                        replace: {  // What new item should replace the existing one (invalidates rename/reimage/replace)
                            newItem: itemId
                        },
                        remove: true,  // Remove the item from inventory at this stage? Invalidates all else
                    },
                    message: "Tagwin's fruit rollup rots away"
                }
            }
        }
    }
}


let settings = {
    stages: {
        1: null,
        2: null
    },
};
let stayOpen = false; // Custom flag to control parent dialog closing

// Create the parent dialog
const parentDialog = new Dialog({
  title: "Parent Dialog",
  content: `
    <!-- Content for the parent dialog goes here -->
  `,
  buttons: {
    stage1: {
      icon: "<i class='fas fa-cog'></i>",
      label: "Configure Stage 1",
      callback: async () => {
        stayOpen = true; // Set the flag to keep the parent dialog open
        openChildDialog(1);
      },
    },
    stage2: {
      icon: "<i class='fas fa-cog'></i>",
      label: "Configure Stage 2",
      callback: async () => {
        stayOpen = true; // Set the flag to keep the parent dialog open
        openChildDialog(2);
      },
    },
    apply: {
      icon: "<i class='fas fa-check'></i>",
      label: "Apply",
      callback: async () => {
        // Handle the Apply button logic here
        console.log("Apply button clicked");
        console.log(settings); // Access the collected settings from child dialogs
        parentDialog.close(); // Close the parent dialog
      },
    },
    cancel: {
      icon: "<i class='fas fa-times'></i>",
      label: "Cancel",
      callback: async () => {
        // Handle the Cancel button logic here
        console.log("Cancel button clicked");
        parentDialog.close(); // Close the parent dialog
      },
    },
  },
  close: () => {
    console.log(settings);
    if (stayOpen) {
      stayOpen = false; // Reset the flag
      parentDialog.render(true); // Re-render the parent dialog to keep it open
    }
  },
  closeOnSubmit: false, // Prevent the parent dialog from closing on submission
  default: 'apply',
}).render(true);


// Function to open a child dialog
// Function to open a child dialog
function openChildDialog(stageInt) {
  const dialog = new Dialog({
    title: "Expiry Options",
    content: `
      <form id="expiry-options-form">
        <div class="form-group">
          <label>Main Option:</label>
          <div class="main-options">
            <fieldset class="option-group">
              <legend>Revise</legend>
              <label>
                <input type="radio" name="main-option" value="revise"> Revise
              </label>
              <div class="settings">
                <label>Prefix:</label>
                <input type="text" name="prefix" class="text-input">
                <label>Suffix:</label>
                <input type="text" name="suffix" class="text-input">
                <label>Image:</label>
                <input type="text" name="img" class="text-input">
              </div>
            </fieldset>
            
            <fieldset class="option-group">
              <legend>Replace</legend>
              <label>
                <input type="radio" name="main-option" value="replace"> Replace
              </label>
              <div class="settings">
                <label>Item:</label>
                <input type="text" name="replacementItem" class="text-input">
              </div>
            </fieldset>

            <fieldset class="option-group">
              <legend>Remove</legend>
              <label>
                <input type="radio" name="main-option" value="remove"> Remove
              </label>
              <div class="settings">
                <!-- Remove settings go here -->
              </div>
            </fieldset>
          </div>
        </div>
      </form>
      <style>
        #savingThrowSelector .dialog-buttons {
            flex-direction: column;
        }
        </style>
    `,
    buttons: {
      apply: {
        icon: "<i class='fas fa-check'></i>",
        label: `Apply`,
        callback: async (html) => {
          // Handle the user's selections here
          const mainOption = html.find('[name="main-option"]:checked').val();
          let stageSettings;

          if (mainOption === "revise") {
            stageSettings = {
              prefix: html.find('[name="prefix"]').val(),
              suffix: html.find('[name="suffix"]').val(),
              img: html.find('[name="img"]').val(),
            };
          } else if (mainOption === "replace") {
            stageSettings = {
              replacementItem: html.find('[name="replacementItem"]').val(),
            };
          } else if (mainOption === "remove") {
            stageSettings = {
              remove: true,
            };
          }

          // Set the settings.stages[stageInt] properties
          console.log("Setting properties")
          settings.stages[stageInt] = stageSettings;

          // Do something with the selected options
          console.log("Main Option:", mainOption);
          console.log("Settings:", stageSettings);

          // Close the dialog
          dialog.close();
        },
      },
      cancel: {
        icon: "<i class='fas fa-times'></i>",
        label: `Cancel`,
      },
    },
    default: 'apply',
  }, {id: "savingThrowSelector"}).render(true);
}

scripts/hooks.js
import { expiryObject, getExpiryActors, getRelevantInventory } from './expiry.js';

Hooks.on("createItem", function(document, _, userId) {
    if(!document.flags?.expiry) {
        return false;
    } else {
        const newExpiryObject = new expiryObject(document, _, userId)
        newExpiryObject.createExpiryObject();
    };
});

Hooks.on("simple-calendar-date-time-change", function(data) {
    const expiryActors = getExpiryActors()
    for(const expiryActor of expiryActors) {
        const affectedItems = getRelevantInventory(expiryActor);
        for(const affectedItem of affectedItems) {
            console.log("updating expiry item")
            const newExpiryObject = new expiryObject(affectedItem, data);
            newExpiryObject.updateExpiryObject();
        };
    };
    Hooks.off("simple-calendar-date-time-change");
});

scripts/expiry.js
export class expiryObject {
    constructor(objectDocument, updateData=null, userId=null) {
        this.objectDocument = objectDocument;
        this.updateData = updateData;
        this.userId = userId;
        this.currentTime = SimpleCalendar.api.timestamp()
    };

    createExpiryObject() {
        let parentExpiry = this.objectDocument.flags.expiry;
        this.objectDocument.update({
            flags: {
                expiry: {
                    instance: {
                        createTime: this.currentTime,
                        createName: this.objectDocument.name,
                        activeStage: 0,
                    }
                }
            }
        });
        console.log("created expiry item")
        Hooks.call('createdExpiry', [this.objectDocument, this.userId])
    };
    
    updateExpiryObject() {
        let expiryData = this.objectDocument.flags.expiry;
        
        // Calculate the time difference in seconds since creation
        const timePassed = this.currentTime - expiryData.instance.createTime;

        // Calculate the number of decay intervals that have passed
        const decaySteps = Math.floor(timePassed / expiryData.decayInterval);

        // Get the array of decay stages from the parentExpiry
        const decayStages = Object.keys(expiryData.decayStages);

        // Determine the current decay stage
        const stageIndex = Math.min(decaySteps, decayStages.length - 1);

        // Update the active stage
        if(!expiryData.instance.activeStage !== stageIndex){
            this.objectDocument.update({
                flags: {
                    expiry: {
                        instance: {
                            activeStage: stageIndex
                        }
                    }
                }
            });
            this._decayExpiryObject(decayStages[stageIndex]);
        };
    };

    _decayExpiryObject(decayStage) {
        // apply decay effects, emit a hook, what have you
        console.log(`decaying item ${this.objectDocument.id}`)
        const decayAction = decayStage.action
        if(decayAction === "revise"){
            this.objectDocument.update({
                img: decayStage.img || this.objectDocument.img,
                name: `${decayStage.prefix}${this.objectDocument.flags.expiry.instance.createName}${decayStage.suffix}` || this.objectDocument.name
            });
        } else if(decayAction === "replace"){
            const parentActor = objectDocument.parent;
            const replacementItem = game.items.get(decayStage.itemId);
            if(await parentActor.createEmbeddedDocuments('Item', [replacementItem.toObject()])){
                this.objectDocument.delete();
            } else {
                console.error(`unable to create replacement object for Expiry ${this.objectDocument.id}, leaving parent object in-place`);
            }
        } else if(decayAction === "remove"){
            this.objectDocument.delete();
        };

        if(decayStage.message){
            ui.notifications.info(decayStage.message);
        };
    };
};

function getRelevantInventory(actorDocument) {
    const affectedItems = actorDocument.items.filter(item => {
      const expiryFlags = item.flags?.expiry;
      if (!expiryFlags) {
        // If the item doesn't have expiry flags, skip it
        return false;
      }
  
      const instance = expiryFlags.instance;
      const decayStages = expiryFlags.decayStages;
  
      if (!instance || !decayStages) {
        // If either instance or decayStages is missing, skip the item
        return false;
      }
  
      // Calculate the entire lifespan of the item
      const lifespanEnd =
        instance.createTime +
        (Object.keys(decayStages).length * expiryFlags.decayInterval);
  
      // Check if the item's createTime is before or equal to the current time,
      // and if the item is not in its final decay state OR if the entire lifespan
      // is greater than or equal to the current time
      return (
        instance.createTime <= this.currentTime &&
        (!isInFinalDecayState(instance.activeStage, decayStages) ||
          lifespanEnd >= this.currentTime)
      );
    });
    return affectedItems;
};
  
  function isInFinalDecayState(activeStage, decayStages) {
    const stages = Object.keys(decayStages);
    return activeStage === stages.length - 1;
};

export function getExpiryActors() {
    // Gets a list of actors who hold expiry-configured items
    return game.actors.filter(actor => actor.items.filter(item => item.flags.expiry));
};

macros/expiryHelp.js
// Function to open a help dialog for all sections
function openHelpDialog() {
    const helpDialog = new Dialog({
      title: "Expiry Item Configuration Help",
      content: `
        <form class="form-horizontal">
          <fieldset>
            <legend>Stage Basics</legend>
            <p>Item lifecycles in Expiry are composed of <b>stages</b> which are progressed through as the game time changes. The lifecycle of an Expiry item may contain up to 3 distinct stages, and each stage may be configured for one action from the list below:</p>
            <ul>
              <li><b>Revise:</b> change the name or image of the item</li>
              <li><b>Replace:</b> replace the item with another item</li>
              <li><b>Remove:</b> remove the item from the player's inventory</li>
            </ul>
            <p>An Expiry lifecycle may mix and match these actions. For instance:</p>
            <ul>
              <li><b>Stage 1:</b> Rename item from 'Banana' to 'Brown Banana' (<i>revise</i>)</li>
              <li><b>Stage 2:</b> Replace item 'Brown Banana' with item 'Rotten Banana,' which you have configured with a status effect to take place when eaten (<i>replace</i>)</li>
              <li><b>Stage 3</b>: Remove the item entirely as it rots to nothing (<i>remove</i>)</li>
            </ul>
            <p>Any stages that are left unconfigured will have no effect on the item; e.g. if the final configured stage changes the name to 'Rotten Banana' and no further action is taken, the item will persist in that state going forward.</p>
          </fieldset>
          
          <fieldset>
            <legend>Stage Decay Rate</legend>
            <p>Stages all follow the same decay rate, set in the main item configuration window. If an item is set to a 12 hour decay rate:</p>
            <ul>
              <li>12 hours after the <b>item is added to an actor's inventory</b> the 1st decay stage will execute</li>
              <li>24 hours later the 2nd action will execute</li>
              <li>36 hours the 3rd action will execute</li>
            </ul>
          </fieldset>

          <fieldset>
            <legend>Reversing Time</legend>
            <p>In its default configuration, Expiry can handle time shifts in reverse as well; that is, in a situation where the GM moves the game clock to an earlier time, decay stages will be reversed.</p>
            <p>This is due to the fact that Expiry lifespans are calculated to a distinct game time, not a relative one. If you add an item to a character's inventory with a 12 hour expiration, then reset the clock to 100 years earlier, it will take 100 years and 12 hours for the item to expire.</p>
            <p>This behavior may be disabled in the module's configuration.</p>
            <p><i>(Note: items that have a <b>remove</b> stage which has already taken place will not reappear)</i></p>
          </fieldset>
          
          <fieldset>
            <legend>Unlimited Variation</legend>
            <p>Individual item stages are limited to a maximum of three, but using the <b>Replace action</b> you may replace the original Expiry-configured item with another Expiry-configured item that has its own distinct configuration. By 'looping' items this way you may achieve infinite variation.</p>
          </fieldset>
          
          <!-- Add more sections as needed -->
        </form>
      `,
      buttons: {
        okay: {
          icon: "<i class='fas fa-check'></i>",
          label: "Okay",
          callback: async () => {
            helpDialog.close(); // Close the help dialog when "Okay" is clicked
          },
        },
      },
      default: 'okay',
    }).render(true);
  }
  
  // Example usage to open the help dialog
  openHelpDialog();

macros/configureExpiryItem.js
let expirySettings = {
  stages: {
      1: null,
      2: null
  },
};
let stayOpen = false; // Custom flag to control parent dialog closing

let itemDropdown = "";
const loadoutsTypes = "gear"

for(const loadoutsType of loadoutsTypes.split(',')){
    itemDropdown +="<option disabled>" + loadoutsType.toUpperCase() + "</option>"
    var itemArray = game.items.filter(item => item.type == loadoutsType).sort((a, b) => a.name.localeCompare(b.name));
    for (let i = 0; i < itemArray.length; i++) {
        var isConfigured
        if(itemArray[i].flags.loadouts){
            if(itemArray[i].flags.loadouts.configured == true){
                isConfigured = "&#x25C9;"
            } else {
                isConfigured = "&#x25CC;"
            }
        } else {
            isConfigured = "&#x25CC;"
        }
        itemDropdown += "<option value='" + itemArray[i].id + "'>" + itemArray[i].name + " " + isConfigured + "</option>";
    }
}

// Create the parent dialog
const parentDialog = new Dialog({
title: "Expiry Item Configuration",
content: `
<form class="form-horizontal">
<fieldset>

  <!-- Form Name -->
  <legend>Item Configuration</legend>
  <!-- Item Dropdown -->
  
  <div class="form-group">
      <label for="selectedItems" style='display:inline-block;'>Select Item(s)</label>
      <select id="selectedItems" name="selectedItems" multiple style='width:58%; margin:4px 1%; display:inline-block;'>` + itemDropdown + `</select>
  </div>
  
  <!-- Decay Time Configuration -->
  <div class="form-group">
      <label for="decayAmount" style='display:inline-block;'>Decay Time:</label>
      <select id="decayAmount" name="decayAmount" style='width:20%; margin:4px 1%; display:inline-block;'>
          <option value="1">1</option>
          <option value="2">2</option>
          <option value="3">3</option>
          <option value="4">4</option>
          <option value="5">5</option>
          <option value="6">6</option>
          <option value="7">7</option>
          <option value="8">8</option>
          <option value="9">9</option>
          <option value="10">10</option>
          <option value="11">11</option>
          <option value="12">12</option>
          <!-- Add more options up to 12 as needed -->
      </select>
      <select id="decayUnit" name="decayUnit" style='width:30%; margin:4px 1%; display:inline-block;'>
          <option value="Minutes">Minutes</option>
          <option value="Hours">Hours</option>
          <option value="Days">Days</option>
          <option value="Years">Years</option>
      </select>
  </div>

</fieldset>
</form>

<style>
  #savingThrowSelector .dialog-buttons {
      flex-direction: column;
  }
</style>
`,

buttons: {
  stage1: {
    icon: "<i class='fas fa-cog'></i>",
    label: "Configure Stage 1",
    callback: async () => {
      stayOpen = true; // Set the flag to keep the parent dialog open
      openChildDialog(1);
    },
  },
  stage2: {
    icon: "<i class='fas fa-cog'></i>",
    label: "Configure Stage 2",
    callback: async () => {
      stayOpen = true; // Set the flag to keep the parent dialog open
      openChildDialog(2);
    },
  },
  reset: {
    icon: "<i class='fas fa-undo'></i>",
    label: "Reset",
    callback: async () => {
      // Handle the Cancel button logic here
      console.log("Cancel button clicked");
      parentDialog.close(); // Close the parent dialog
    },
  },
  help: {
    icon: "<i class='fas fa-info-circle'></i>",
    label: "Help",
    callback: async () => {
      // Handle the Cancel button logic here
      console.log("Help button clicked");
      stayOpen = true; // Set the flag to keep the parent dialog open
      openHelpDialog(); // Close the parent dialog
    },
  },
  apply: {
    icon: "<i class='fas fa-check'></i>",
    label: "Apply",
    callback: async () => {
      // Handle the Apply button logic here
      console.log("Apply button clicked");
      console.log(settings); // Access the collected settings from child dialogs
      parentDialog.close(); // Close the parent dialog
    },
  },
  cancel: {
    icon: "<i class='fas fa-times'></i>",
    label: "Cancel",
    callback: async () => {
      // Handle the Cancel button logic here
      console.log("Cancel button clicked");
      parentDialog.close(); // Close the parent dialog
    },
  },
},
close: () => {
  console.log(settings);
  if (stayOpen) {
    stayOpen = false; // Reset the flag
    parentDialog.render(true); // Re-render the parent dialog to keep it open
  }
},
closeOnSubmit: false // Prevent the parent dialog from closing on submission
}, {id: "savingThrowSelector"}).render(true);

// Function to open a child dialog
function openChildDialog(stageInt) {
const dialog = new Dialog({
  title: `Expiry Stage ${stageInt} Settings`,
  content: `
    <form id="expiry-options-form">
      <div class="form-group">
        <label>Action:</label>
        <div class="main-options">

          <fieldset class="option-group">
            <legend>Revise</legend>
            <label>
              <input type="radio" name="main-option" value="revise"> Revise
            </label>
            <div class="settings">
              <label>Prefix:</label>
              <input type="text" name="prefix" class="text-input">
              <label>Suffix:</label>
              <input type="text" name="suffix" class="text-input">
              <label>Image:</label>
              <input type="text" name="img" class="text-input">
            </div>
          </fieldset>
          
          <fieldset class="option-group">
            <legend>Replace</legend>
            <label>
              <input type="radio" name="main-option" value="replace"> Replace
            </label>
            <div class="settings">
              <label>Item:</label>
              <input type="text" name="replacementItem" class="text-input">
            </div>
          </fieldset>

          <fieldset class="option-group">
            <legend>Remove</legend>
            <label>
              <input type="radio" name="main-option" value="remove"> Remove
            </label>
            <div class="settings">
              <!-- Remove settings go here -->
            </div>
          </fieldset>
        </div>
      </div>

      <div class="form-group">
      <label>Notification (optional):</label>
      <div class="notification-options">
          <fieldset class="option-group">
            <legend>UI Message</legend>
            <div class="settings">
              <label>Message:</label>
                <input type="text" name="message" class="text-input">
              <label>Send to:</label>
                <br>
                <input type="checkbox" name="console"> Console
              </label>
              <label>
                <input type="checkbox" name="chat"> Chat
              </label>
              <label>
                <input type="checkbox" name="ui"> UI
              </label>
            </div>
          </fieldset>
        </div>
      </div>

    </form>
  `,
  buttons: {
    apply: {
      icon: "<i class='fas fa-check'></i>",
      label: `Apply`,
      callback: async (html) => {
        // Handle the user's selections here
        const mainOption = html.find('[name="main-option"]:checked').val();
        let stageSettings;

        if (mainOption === "revise") {
          stageSettings = {
            action: 'revise',
            prefix: html.find('[name="prefix"]').val(),
            suffix: html.find('[name="suffix"]').val(),
            img: html.find('[name="img"]').val()
          };
        } else if (mainOption === "replace") {
          stageSettings = {
            action: 'replace',
            replacementItem: html.find('[name="replacementItem"]').val()
          };
        } else if (mainOption === "remove") {
          stageSettings = {
            action: 'remove'
          };
        };

        // Get notification settings
        const decayMessage = html.find('[name="message"]').val();
        const consoleChecked = html.find('[name="console"]').prop('checked');
        const chatChecked = html.find('[name="chat"]').prop('checked');
        const uiChecked = html.find('[name="ui"]').prop('checked');

        // Include the checkbox values in the stageSettings
        stageSettings.notifications = {
          message: decayMessage,
          to_console: consoleChecked,
          to_chat: chatChecked,
          to_ui: uiChecked
        }

        // Set the settings.stages[stageInt] properties
        console.log("Setting properties")
        expirySettings.stages[stageInt] = stageSettings;

        // Do something with the selected options
        console.log("Main Option:", mainOption);
        console.log("Settings:", stageSettings);

        // Close the dialog
        dialog.close();
        },
      },
      cancel: {
        icon: "<i class='fas fa-times'></i>",
        label: `Cancel`,
      },
    },
    default: 'apply',
  }).render(true);
};

// Function to open a help dialog for all sections
function openHelpDialog() {
  const helpDialog = new Dialog({
    title: "Expiry Item Configuration Help",
    content: `
    <form class="form-horizontal">
        <fieldset>
        <legend>Stage Basics</legend>
        <p>Item lifecycles in Expiry are composed of <b>stages</b> which are progressed through as the game time changes. The lifecycle of an Expiry item may contain up to 3 distinct stages, and each stage may be configured for one action from the list below:</p>
        <ul>
            <li><b>Revise:</b> change the name or image of the item</li>
            <li><b>Replace:</b> replace the item with another item</li>
            <li><b>Remove:</b> remove the item from the player's inventory</li>
        </ul>
        <p>An Expiry lifecycle may mix and match these actions. For instance:</p>
        <ul>
            <li><b>Stage 1:</b> Rename item from 'Banana' to 'Brown Banana' (<i>revise</i>)</li>
            <li><b>Stage 1 (alt):</b> Replace item 'Brown Banana' with item 'Rotten Banana,' which you have configured with a status effect to take place when eaten (<i>replace</i>)</li>
            <li><b>Stage 2</b>: Remove the item entirely as it rots to nothing (<i>remove</i>)</li>
        </ul>
        <p>Any stages that are left unconfigured will have no effect on the item; e.g. if the final configured stage changes the name to 'Rotten Banana' and no further action is taken, the item will persist in that state going forward.</p>
        </fieldset>
        
        <fieldset>
        <legend>Stage Decay Rate</legend>
        <p>Stages all follow the same decay rate, set in the main item configuration window. If an item is set to a 12 hour decay rate:</p>
        <ul>
            <li>12 hours <b>after an instance of the item is added to an actor's inventory</b> the 1st decay stage will execute</li>
            <li>24 hours later the 2nd action will execute</li>
            <li>36 hours the 3rd action will execute</li>
        </ul>
        </fieldset>

        <fieldset>
        <legend>Reversing Time</legend>
        <p>In its default configuration, Expiry can handle time shifts in reverse as well; that is, in a situation where the GM moves the game clock to an earlier time, decay stages will be reversed.</p>
        <p>This is due to the fact that Expiry lifespans are calculated to a distinct game time, not a relative one. If you add an item to a character's inventory with a 12 hour expiration, then reset the clock to 100 years earlier, it will take 100 years and 12 hours for the item to expire.</p>
        <p>Time reversal behavior may be disabled in the module's configuration.</p>
        <p><i>(Note: items that have a <b>remove</b> stage which has already taken place will not reappear)</i></p>
        </fieldset>
        
        <fieldset>
        <legend>Unlimited Variation</legend>
        <p>Individual item stages are limited to a maximum of three, but using the <b>Replace action</b> you may replace the original Expiry-configured item with another Expiry-configured item that has its own distinct configuration. By 'looping' items this way you can get as detailed as you desire.</p>
        </fieldset>
        
        <!-- Add more sections as needed -->
    </form>
    `,
    buttons: {
    okay: {
        icon: "<i class='fas fa-check'></i>",
        label: "Okay",
        callback: async () => {
        helpDialog.close(); // Close the help dialog when "Okay" is clicked
        },
    },
    },
    default: 'okay',
  }).render(true);
};
